server/asset/eth/tokener.go

```diff
@@ -0,0 +1,150 @@
+	erc20v0 "decred.org/dcrdex/dex/networks/erc20/contracts/v0"
+	dexeth "decred.org/dcrdex/dex/networks/eth"
+	swapv0 "decred.org/dcrdex/dex/networks/eth/contracts/v0"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// swapContract is a generic source of swap contract data.
+type swapContract interface {
+	Swap(context.Context, [32]byte) (*dexeth.SwapState, error)
+}
+
+// tokenAddresser exposes the TokenAddress method of a token swap contract.
+type tokenAddresser interface {
+	TokenAddress(*bind.CallOpts) (common.Address, error)
+}
+
+// erc2Contract exposes methods of a token's ERC20 contract.
+type erc20Contract interface {
+	BalanceOf(*bind.CallOpts, common.Address) (*big.Int, error)
+}
+
+// tokener is a contract data manager for a token.
+type tokener struct {
+	*registeredToken
+	swapContract
+	tokenAddresser
+	erc20Contract
+	ver                     uint32
+	contractAddr, tokenAddr common.Address
+}
+
+// newTokener is a constructor for a tokener.
+func newTokener(ctx context.Context, assetID uint32, net dex.Network, be bind.ContractBackend) (*tokener, error) {
+	token, netToken, swapContract, err := networkToken(assetID, net)
+	if err != nil {
+		return nil, err
+	}
+
+	if token.ver != 0 {
+		return nil, fmt.Errorf("only version 0 contracts supported")
+	}
+
+	es, err := erc20v0.NewERC20Swap(swapContract.Address, be)
+	if err != nil {
+		return nil, err
+	}
+
+	erc20, err := erc20.NewIERC20(netToken.Address, be)
+	if err != nil {
+		return nil, err
+	}
+
+	tkn := &tokener{
+		ver:             0,
+		registeredToken: token,
+		swapContract:    &swapSourceV0{es},
+		tokenAddresser:  es,
+		erc20Contract:   erc20,
+		contractAddr:    swapContract.Address,
+		tokenAddr:       netToken.Address,
+	}
+
+	if boundAddr, err := tkn.tokenAddress(ctx); err != nil {
+		return nil, fmt.Errorf("error retrieving bound address for %s version %d contract: %w",
+			token.Name, token.ver, err)
+	} else if boundAddr != netToken.Address {
+		return nil, fmt.Errorf("wrong bound address for %s version %d contract. wanted %s, got %s",
+			token.Name, token.ver, netToken.Address, boundAddr)
+	}
+
+	return tkn, nil
+}
+
+// transferred calculates the value transferred using the token contract's
+// transfer method.
+func (t *tokener) transferred(txData []byte) *big.Int {
+	_, out, err := erc20.ParseTransferData(txData)
+	if err != nil {
+		return nil
+	}
+	return out
+}
+
+// swapped calculates the value sent to the swap contracts initiate method.
+func (t *tokener) swapped(txData []byte) *big.Int {
+	inits, err := dexeth.ParseInitiateData(txData, t.ver)
+	if err != nil {
+		return nil
+	}
+	var v uint64
+	for _, init := range inits {
+		v += init.Value
+	}
+	return dexeth.GweiToWei(v)
+}
+
+// tokenAddress fetches the token address bound to the swap contract.
+func (t *tokener) tokenAddress(ctx context.Context) (common.Address, error) {
+	return t.tokenAddresser.TokenAddress(readOnlyCallOpts(ctx))
```

Comment:
`tokenAddresser` can be omitted like you did in `balanceOf` with `erc20Contract`